### ISSUE TITLE
Bug on react native polyfill for Object.assign

### ISSUE THREAD

**Issue by tychota** [apollostack/react-apollo#95]
cc @jbaxleyiii 
# Short description

Running it on RN, I get:
```
 One of the sources to assign has an enumerable key on the prototype chain. This is an edge case that we do not support. This error is a performance optimization and not spec compliant.
```
Seems related to recent  #94 

# Package versions

|  Package | Version |
| ------------- | ------------- |
| apollo-client | 0.3.27 |
| graphql-tag | 0.1.9 |
| react | 15.2.0 |
| react-apollo | 0.3.14 |
| react-native | 0.29.0 |

# Context

<details> 
  <summary>**BooksListPage.js**</summary>

```js
import React from 'react';
import { View } from 'react-native';
import { connect } from 'react-apollo';
import gql from 'graphql-tag';

import BooksList from 'Gooz/js/components/BooksList';

const style = {
  container: {
    marginTop: 64,
  },
};

type Props = {
  books: Array<Book>;
};

const BooksListPage = (props: Props) => (
  <View style={style.container}>
    <BooksList books={props.books} />
  </View>
);

function mapQueriesToProps() {
  return {
    category: {
      query: gql`
        query getBooksList() {
          books: allBooks(first: 10) {
            book {
              title
              id
            }
          }
        }
      `,
      forceFetch: false,
      returnPartialData: true,
    },
  };
}

const mapStateToProps = ({ routes }) => ({ routes });

export default connect(mapStateToProps, mapQueriesToProps)(BooksListPage);
```
</details>

<details> 
  <summary>**BookList.js**</summary>

```js
import React, { Component } from 'react';
import { Actions } from 'react-native-router-flux';

import BookCell from 'Gooz/js/components/BookCell';
import ListView from 'Gooz/js/components/common/ListView';

type Props = {
  books: Array<Book>;
  renderEmptyList?: () => ReactElement<*>;
};
type State = {
  books: Array<Book>;
};

type RenderRow = (book: Book) => ReactElement<{book: Book, onPress: () => void}>;

class BooksList extends Component {
  props: Props;
  state: State;
  _innerRef: ?ListView;

  constructor(props: Props): void {
    super(props);
    this.state = {
      books: props.books,
    };
    this._innerRef = null;
  }

  componentWillReceiveProps(nextProps: Props): void {
    if (nextProps.books !== this.props.books) {
      this.setState({
        books: nextProps.books,
      });
    }
  }

  renderRow: RenderRow = (book) => <BookCell onPress={() => this.openBook(book)} book={book} />;

  render() {
    return (
      <ListView
        ref={this.storeInnerRef}
        data={this.state.books}
        renderRow={this.renderRow}
        {...this.props}
        renderEmptyList={this.renderEmptyList}
      />
    );
  }

  renderEmptyList: () => ?ReactElement<*> = () => {
    const { renderEmptyList } = this.props;
    return renderEmptyList && renderEmptyList();
  }

  openBook: (book: Book) => void = (book) => {
    Actions.bookDetail(book.id);
  }

  storeInnerRef: (ref: ?ListView) => void = (ref) => {
    this._innerRef = ref;
  }
}

export default BooksList;
```
</details>

# Bug digging

<details> 
  <summary>Copy Paste of the polyfill used by RN</summary>
```js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * This pipes all of our console logging functions to native logging so that
 * JavaScript errors in required modules show up in Xcode via NSLog.
 *
 * @provides Object.es6
 * @polyfill
 */

// WARNING: This is an optimized version that fails on hasOwnProperty checks
// and non objects. It's not spec-compliant. It's a perf optimization.
/* eslint strict:0 */
Object.assign = function(target, sources) {
  if (__DEV__) {
    if (target == null) {
      throw new TypeError('Object.assign target cannot be null or undefined');
    }
    if (typeof target !== 'object' && typeof target !== 'function') {
      throw new TypeError(
        'In this environment the target of assign MUST be an object.' +
        'This error is a performance optimization and not spec compliant.'
      );
    }
  }

  for (var nextIndex = 1; nextIndex < arguments.length; nextIndex++) {
    var nextSource = arguments[nextIndex];
    if (nextSource == null) {
      continue;
    }

    if (__DEV__) {
      if (typeof nextSource !== 'object' &&
          typeof nextSource !== 'function') {
        throw new TypeError(
          'In this environment the sources for assign MUST be an object.' +
          'This error is a performance optimization and not spec compliant.'
        );
      }
    }

    // We don't currently support accessors nor proxies. Therefore this
    // copy cannot throw. If we ever supported this then we must handle
    // exceptions and side-effects.

    for (var key in nextSource) {
      if (__DEV__) {
        var hasOwnProperty = Object.prototype.hasOwnProperty;
        if (!hasOwnProperty.call(nextSource, key)) {
          throw new TypeError(
            'One of the sources for assign has an enumerable key on the ' +
            'prototype chain. This is an edge case that we do not support. ' +
            'This error is a performance optimization and not spec compliant.'
          );
        }
      }
      target[key] = nextSource[key];
    }
  }

  return target;
};
```

</details>

putting a chrome breakpoint I get:

<img width="1123" alt="screen shot 2016-07-06 at 22 54 31" src="https://cloud.githubusercontent.com/assets/13785185/16634128/aac995b2-43cc-11e6-9f3a-bedb85874556.png">

I've no time till saturday to PR this.
Hope you have enough details to do the hard job ;)

**Comment by tychota** [comment]
Till this resolved, going back to 0.3.13 of react-apollo fix this.

**Comment by sync** [comment]
same problem here, going back to 0.3.13 did fix it
